src/tracer.cc

```diff
@@ -580,45 +583,37 @@ TraceManager::Trace::StartSpan(
   options.start_steady_time =
       otel_common::SteadyTimestamp{std::chrono::nanoseconds{raw_timestamp_ns}};
 
-  // If the new span is a child span, we need to retrieve its parent from
-  // the context and provide it through StartSpanOptions to the child span
-  if (!parent_span_key.empty() && otel_context_.HasKey(parent_span_key)) {
-    auto parent_span = opentelemetry::nostd::get<
-        opentelemetry::nostd::shared_ptr<otel_trace_api::Span>>(
-        otel_context_.GetValue(parent_span_key));
-    options.parent = parent_span->GetContext();
+  // If the new span is a child span, we need to retrieve its parent and
+  // provide it through StartSpanOptions to the child span
+  if (span_stacks_.find(trace_id) != span_stacks_.end() &&
+      !span_stacks_[trace_id]->empty()) {
```

Comment:
Won't this be entered twice if the caller: start span -> end span -> start span? Will this mistakenly treat the trace to "have" a parent?

Reply:
Every `end span` event pops span from the stack, so it doesn't exist anymore. 
The trace frontend currently controls root span creation, request spans and compute spans. The passed `trace_id` in this method obides the logic, explained here: https://github.com/triton-inference-server/server/pull/7390/files#diff-c05667609085c993fd282ff18d0429945c44492b95c0d01fcbeec6946fda2db2R525-R534